### PR TITLE
Duration created with no parts will have a default seconds part eqaul to 0

### LINKED ISSUE #31302: `0 % <ActiveSupport::Duration>` returns unexpected result
### Steps to reproduce

From the Rails console:
`>> a = 0 % 1.second`
`>> a`
`#>`

### Expected behavior
`a` should report its existence as `0`.

### Actual behavior
`a` returns nothing at all: No `0`, no `nil`, no indication of what it is. However,

`>> a.zero?`
`#> true`

### System configuration
**Rails version**: 5.1.4

**Ruby version**: 2.3.5

### PR DESCRIPTION
**Pull request by kinnrot**
Added a default value to seconds part with parts are empty.
Fixes #31302

### CODE REVIEW

**Comment by pixeltrix** [comment]
@kinnrot  thanks for the PR

It seems like #31302 is really an issue with `inspect` - doing maths with the object works fine, e.g.

```
>> (10 % 5.seconds) == 0
=> true
```

Given that I think we just need to add `return "0 seconds" if parts.empty?` to the `inspect` method.  Also in the test, rather than building a object directly can you use the modulo operation to get the empty duration, e.g:

```
assert_equal "0 seconds", (10 % 5.seconds).inspect
```

**Comment by pixeltrix** [comment]
Actually, whilst you're making changes can you reject zero parts in `initialize`, e.g:

``` ruby
def initialize(value, parts) #:nodoc:
  @value, @parts = value, parts.to_h
  @parts.default = 0
  @parts.reject! { |k, v| v.zero? }
end
```

and another line to `test_inspect`:

``` ruby
assert_equal "10 minutes", (10.minutes + 0.seconds).inspect
```

Otherwise we leak these zero parts when combining durations, e.g:

```
>> 10.minutes + 0.seconds
=> 10 minutes, and 0 seconds
```

Thanks! 👍

**Comment by eugeneius** [comment]
If `initialize` rejects empty parts, we can also ✂️ this line:

https://github.com/rails/rails/blob/067fc779c4560fff4812614a2f78f9248f3e55f8/activesupport/lib/active_support/duration.rb#L197

File: activesupport/lib/active_support/duration.rb [hunk 1]
```diff
@@ -370,6 +370,8 @@ def ago(time = ::Time.current)
     alias :before :ago
 
     def inspect #:nodoc:
+      return "0 seconds" if parts.empty?
+
```
Comment (eugeneius) [inline activesupport/lib/active_support/duration.rb]: Rubocop is complaining about the trailing spaces on this line, can you remove them? ✂️

File: activesupport/test/core_ext/duration_test.rb [hunk 2]
```diff
@@ -71,6 +71,9 @@ def test_inspect
     assert_equal "7 days",                          7.days.inspect
     assert_equal "1 week",                          1.week.inspect
     assert_equal "2 weeks",                         1.fortnight.inspect
+    assert_equal "0 seconds",                       ActiveSupport::Duration.new(0, []).inspect
```
Comment (eugeneius) [inline activesupport/test/core_ext/duration_test.rb]: `ActiveSupport::Duration.new` isn't part of the public API, so we don't need to add tests for it; the one below using `%` is sufficient.

File: activesupport/lib/active_support/duration.rb (hunk 1 above)
Comment (kinnrot) [reply activesupport/lib/active_support/duration.rb]: done

File: activesupport/test/core_ext/duration_test.rb (hunk 2 above)
Comment (kinnrot) [reply activesupport/test/core_ext/duration_test.rb]: done

File: activesupport/lib/active_support/duration.rb [hunk 3]
```diff
@@ -211,6 +210,7 @@ def calculate_total_seconds(parts)
     def initialize(value, parts) #:nodoc:
       @value, @parts = value, parts.to_h
       @parts.default = 0
+      parts.reject! { |k, v| v.zero? }
```
Comment (eugeneius) [inline activesupport/lib/active_support/duration.rb]: This should be modifying `@parts` (the object's state), not `parts` (the initializer argument).

If `parts` is a hash then they're the same object and it doesn't matter, but sometimes it's an array-of-arrays:

https://github.com/rails/rails/blob/v5.1.4/activesupport/lib/active_support/duration.rb#L153

Which leads to this strange behaviour:

```ruby
0.minutes
=> 0 minutes
0.minutes + 0.minutes
=> 0 seconds
```

**Comment by kinnrot** [comment]
Hi guys, can you merge ?

**Pull request merged**